### DISCUSSION TITLE
Having Troubles with Blocks and Text Tags

### DISCUSSION THREAD

**Discussion by idontcompletethings** [mozman/ezdxf#961] [Q&A]
Hi There,

I'd consider myself very much a beginner; so I very much think I am missing something - but I cannot seem to find any answers so far!

My goal is to have a 'blocksdoc' dxf file that contains a number of blocks in a single place. This document is created and saved from within Autocad 2024 to r2018. Each of these blocks has one or more Tags of single line text (no multiline etc.).

I have been able to import these blocks into my ezdxf generated dxf file (test.dxf), and I can place the blocks, however the Tag doesn't seem to exist' and looking at the 'Launcher' tool, there is no 'attdef' section under my block.

However, if I create the block using ezdxf (say for example a file named ezdxfblock), and follow the same process of importing this dxf file into to my test.dxf file, and use the Launcher tool, I can see the 'attdef'.

Note: Even with it there I am still uncertain how to actually populate it.

If someone could guide me on the first issue of not having the attdef present, when the block has been generated in Autocad, and also assist in how to access the Tag, I would be most grateful!
Thanks in advance.

**Comment by mozman** [comment]
Try to save the DXF file as DXF R2013 from AutoCAD. If this doesn't help you have to post your code and the DXF file from AutoCAD as a zip archive.

**Comment by idontcompletethings** [comment]
Hi there, thanks for your suggestion. I have created the blocks drawing as r2013 dxf, but unfortunately nothing has changed.

Please see the attached as requested. I have excluded code where I am trying to 'access' the Tag (this is because I am not 100% sure how to do it, and a lot of things I tried kept giving an error).
However, from the 'Launcher' tool, to look at the DXF structure, I still do not see the 'ATTDEF' for the blocks.
Appreciate the assistance. Thanks.

[block_testing.zip](https://github.com/mozman/ezdxf/files/13328336/block_testing.zip)

**Comment by mozman** [comment]
Hi!

1. Do not use the `Importer` add-on, it's outdated and replaced by the `ezdxf.xref` module
2. When using the `Importer` you have to call `finalize()` on the importer object
3. Unfortunately nothing of these tips is helping in this case
    - AutoCAD added an extension dictionary to the `ATTDEF` entity, which `ezdxf` cannot copy because some unknown/undocumented entities are included

The only solution that works is to remove all extension dictionaries from the entities in the block.

```Python
from pathlib import Path
import ezdxf
from ezdxf import colors
from ezdxf import xref

CWD = Path(r"C:\Users\mozman\Desktop\Now\ezdxf\961")
# Read the file where my blocks live
source_doc = ezdxf.readfile(CWD / "blocksdoc2013.dxf")
# Create a new DXF document.
target_doc = ezdxf.new(dxfversion=source_doc.dxfversion, units=source_doc.units)

# Load blocks by the xref module
block_loader = xref.Loader(source_doc, target_doc)
for block_name in ["FUSEDXF", 'CIRCUIT_BREAKERDXF']:
    block = source_doc.blocks.get(block_name)
    for entity in block:
        entity.discard_extension_dict()
    block_loader.load_block_layout(block)
# execute import process
block_loader.execute()

target_doc.layers.add("Layer1", color=colors.YELLOW, lineweight=30)
msp = target_doc.modelspace()
msp.add_blockref('FUSEDXF', (0,0), dxfattribs={"layer": "Layer1"})
target_doc.saveas(CWD / "output.dxf")
```
This works, but also removes some unknown data!!!

![image](https://github.com/mozman/ezdxf/assets/4230966/df362045-ed06-4479-8828-604a654ada62)

The best solution is always to use the source document as template:
```Python
from pathlib import Path
import ezdxf
from ezdxf import colors

CWD = Path(r"C:\Users\mozman\Desktop\Now\ezdxf\961")
# load the source doc as template
doc = ezdxf.readfile(CWD / "blocksdoc2013.dxf")

doc.layers.add("Layer1", color=colors.YELLOW, lineweight=30)
msp = doc.modelspace()
msp.add_blockref('FUSEDXF', (0,0), dxfattribs={"layer": "Layer1"})
# export as new DXF document
doc.saveas(CWD / "output.dxf")
```

Composing new DXF documents from multiple DXF files often fail because newer AutoCAD versions use many undocumented entities and features which can be preserved by `ezdxf` but not copied.

**Reply by idontcompletethings** [reply]
Thanks, I will take a look at this and try to digest it, and see how it works.

However, whilst it may be there, it's not obvious to me (as I say, I'm very new to trying to program!)

But with what you have added, where would I define the value of the tag in the block? And furthermore, if I have multiple instances of the same block (e.g. 5 'FUSEDXF', how would I identify them individually and then the tag associated (as values may be different for each instance)?

Thanks

**Reply by mozman** [reply]
There is a tutorial for using blocks and attributes: https://ezdxf.mozman.at/docs/tutorials/blocks.html

**Reply by idontcompletethings** [reply]
Thanks for the link! I have read this over and over, but being a newbie, there is something that I am not following and can't piece together! 
I'll continue to try, but finding further resources seems to be a little difficult, and that's what I am needing really. But I'll keep at it!

Even if I don't work out how to edit the text in a tag, I can work around it all by just adding a text field each time. But I wanted to build it up, how I would do manually and that would be by using blocks, and I had planned to make the 'blocksdoc' a library of symbols and the associated text, so it can be populated with whatever information!

**Comment by mozman** [comment]
Deep down the rabbit hole I found the culprit `ACDB_MTEXTATTRIBUTEOBJECTCONTEXTDATA_CLASS`

- Block `300` "circuit_breakerdxf contains an ...
- `ATTDEF` `2F9` has an ...
- extension dict `2FA`, has an entry ( "AcDbContextDataManager", `2FB`)
- `DICTIONARY` `2FB`, has an  entry ("ACDB_ANNOTATIONSCALES", `2FC`)
- `DICTIONARY` `2FC`, has an entry ("*A1", `2FD`)
- `ACDB_MTEXTATTRIBUTEOBJECTCONTEXTDATA_CLASS` `2FD`

**Reply by idontcompletethings** [reply]
I'll be completely honest; I'm not sure what that all means! 
I know I left the 'example text' of the tag as `...A` but removed it from the Fuse block, but seem to have left it on the circuit_breaker block, if that helps clear anything up?

What was the outcome of the rabbit hole, is that pointing to a specific place i need to look?
Thanks
